pallets/core_mods/src/storage_reader_tests.rs

```diff
@@ -0,0 +1,198 @@
+        Error, MetaStorageReader,
+    },
+    output::RawStorageValue,
+    params::Params,
+};
+
+pallet_evm_precompile_storage_reader::impl_pallet_storage_metadata_provider!(
+    for Test:
+        "DIDModule" => DIDModule
+);
+
+#[derive(Eq, PartialEq, Debug)]
+enum ErrorWrapper {
+    Execution(ExitError),
+    Decoding(codec::Error),
+}
+
+impl From<ExitError> for ErrorWrapper {
+    fn from(err: ExitError) -> Self {
+        Self::Execution(err)
+    }
+}
+
+impl From<codec::Error> for ErrorWrapper {
+    fn from(err: codec::Error) -> Self {
+        Self::Decoding(err)
+    }
+}
+
+static DUMMY_CTX: &'static evm::Context = &evm::Context {
+    address: H160([0; 20]),
+    caller: H160([0; 20]),
+    apparent_value: U256([32; 4]),
+};
+
+macro_rules! assert_returned_value {
+    ($expr: expr, $val: expr) => {{
+        assert_eq!(
+            $expr.map_err(ErrorWrapper::Execution).and_then(
+                |PrecompileOutput {
+                     output,
+                     exit_status,
+                     ..
+                 }| {
+                    assert_eq!(exit_status, ExitSucceed::Returned);
+
+                    let raw = RawStorageValue::decode_from_bytes(&output[..]);
+                    raw.into_item()
+                        .map(|bytes| {
+                            Decode::decode(&mut &bytes[..]).map_err(ErrorWrapper::Decoding)
+                        })
+                        .transpose()
+                }
+            ),
+            $val
+        );
+    }};
+}
+
+#[test]
+fn invalid_input() {
+    ext().execute_with(|| {
+        let input = Input::new("Pallet", "Version", NoKey, Params::None);
+        assert_returned_value!(
+            MetaStorageReader::<Test>::execute(&input.encode(), Some(10_000_000), DUMMY_CTX),
+            Err::<Option<StoredDidDetails<Test>>, _>(ExitError::from(Error::MemberNotFound).into())
+        );
+
+        let input = Input::new("DIDModule", "Field", NoKey, Params::None);
+        assert_returned_value!(
+            MetaStorageReader::<Test>::execute(&input.encode(), Some(10_000_000), DUMMY_CTX),
+            Err::<Option<StoredDidDetails<Test>>, _>(ExitError::from(Error::MemberNotFound).into())
+        );
+
+        let input = Input::new("DIDModule", "DidControllers", NoKey, Params::None);
+        assert_returned_value!(
+            MetaStorageReader::<Test>::execute(&input.encode(), Some(10_000_000), DUMMY_CTX),
+            Err::<Option<StoredDidDetails<Test>>, _>(ExitError::from(Error::InvalidKey).into())
+        );
+
+        let did = Did([2u8; 32]);
+        let details: StoredDidDetails<Test> =
+            WithNonce::new_with_nonce(OnChainDidDetails::new(IncId::from(1u32), 2, 3), 5).into();
+        Dids::insert(did, &details);
+
+        let input = Input::new("DIDModule", "Dids", MapKey::new(did), Params::None);
+        assert_returned_value!(
+            MetaStorageReader::<Test>::execute(&input.encode(), Some(10_000_000), DUMMY_CTX),
```

Comment:
Why does this fail?

Reply:
Attempt to encode invalid entity (`Did` instead of `StoredDidDetails `)

Reply:
Then why does [this](https://github.com/docknetwork/dock-substrate/pull/118/files#diff-16572ab34bcf31e1999df400d79ba58cd6496d2fddfcfbf0c0f0f8ba6d07f763R139) succeed? Before both statements, you have a DID being inserted.

Reply:
[Here](https://github.com/docknetwork/dock-substrate/pull/118/files#diff-16572ab34bcf31e1999df400d79ba58cd6496d2fddfcfbf0c0f0f8ba6d07f763R109) there's an attempt to decode details as `Did`, that's why it fails.
In a [successful test](https://github.com/docknetwork/dock-substrate/pull/118/files#diff-16572ab34bcf31e1999df400d79ba58cd6496d2fddfcfbf0c0f0f8ba6d07f763R142) it decodes details as details, that's why it succeeds.
> Before both statements, you have a DID being inserted.

In both tests firstly we insert `DidDetails` in a map `Did` -> `DidDetails`. However, in the failed test we attempt to decode inserted did details as `Did` while in the successful test we decode it correctly.

Reply:
Yes, that's the reason.

Reply:
Ok